### PR TITLE
fix: do not check existing controller for register --replace

### PR DESCRIPTION
**Pull request by wallyworld**
TL;DR: When registering a new user, if  `--replace` is specified, do not do the initial check for an existing controller with the same address. This means it's possible to burn a token, but the overall solution is still better that what used to be the case where a token could be burnt every time.

https://github.com/juju/juju/pull/18079 contained a fix to deal with an error when registering a controller where the user is already logged in. Part of the fix was to recognise that an initial check for there being an existing controller with the same endpoints was only performed if a public hostname was specified for the controller. This check is applicable for non public controllers also.

The problem is that when `--replace` is specified, the "is there an existing controller with the same endpoint" check would "correctly" fail. This wasn't noticed previously because the check was only done for public endpoints. So now we only do the check if the user hasn't specified `--replace`. The public vs non-public checks have been tuned to have subtly different error messages so introduce separate methods.

## QA steps

bootstrap
juju add-user bob
juju grant bob read controller

in another terminal
export JUJU_DATA=/path/to/bob
copy controllers.yaml to $JUJU_DATA
edit controllers.yaml and change UUID
juju register ...
ERROR A controller with the same API address has already\n been registered ...
juju register --replace ...

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2073741

**Jira card:** [JUJU-7287](https://warthogs.atlassian.net/browse/JUJU-7287)

### CODE REVIEW

**Comment by sinanawad** [comment]
@wallyworld I'm confused. The sum of the two PRs fix the same bug? as it doesn't seem a regression.
Please also update the bug with target release.

**Comment by wallyworld** [comment]
> @wallyworld I'm confused. The sum of the two PRs fix the same bug? as it doesn't seem a regression. Please also update the bug with target release.

The bug should have been marked as fix committed with the first PR and assigned to 3.5.5. This PR provides some tweaks to the error messages to improve the UX since the first PR got the messages a bit wrong. The plan was to update the bug accordingly when this lands.

File: cmd/juju/controller/register.go [hunk 1]
```diff
@@ -811,31 +844,56 @@ func genAlreadyRegisteredError(controller, user string) error {
 }
 
 var alreadyRegisteredMessageTokenStillValidT = template.Must(template.New("").Parse(`
-This controller has already been registered on this client as "{{.ControllerName}}".
-To login as user "{{.UserName}}" run 'juju login -u {{.UserName}} -c {{.ControllerName}}'.
+A controller with the same API address has already been registered on this client as "{{.ControllerName}}".
```
Comment (hpidcock) [inline cmd/juju/controller/register.go:847]:
```suggestion
A controller with the same address has already been registered on this client as "{{.ControllerName}}".
```

File: cmd/juju/controller/register.go [hunk 2]
```diff
@@ -811,31 +844,56 @@ func genAlreadyRegisteredError(controller, user string) error {
...
+			LoggedInUserName string
+		}{ControllerName: controller, UserName: user, LoggedInUserName: loggedInUser},
+	); err != nil {
+		return err
+	}
+	return errors.New(buf.String())
+}
+
+var alreadyRegisteredPublicMessageTokenStillValidT = template.Must(template.New("").Parse(`
+A controller with the same API address has already been registered on this client as "{{.ControllerName}}".
```
Comment (hpidcock) [inline cmd/juju/controller/register.go:872]:
```suggestion
A controller with the same address has already been registered on this client as "{{.ControllerName}}".
```

**Review by hpidcock — APPROVED** [review]
